### PR TITLE
[GPU] Keep RMS normalization nodes in FP32 to prevent NaN from FP16 overflow

### PR DESCRIPTION
**Pull request by zhanmyz**
### Details:
  - Keep RMS normalization nodes in FP32 to prevent NaN caused by FP16 overflow in models with large intermediate activations (ParlerTTS, flan-T5).
### Description of the issue (symptom, root-cause, how it was resolved)
  - **Symptom**:
    - ParlerTTS Mini text_encoder (T5-based, `google/flan-t5-base`) produced all-NaN outputs on GPU FP16 (`encoder_last_hidden_state`, `logits`).
    - GPU FP32 and CPU modes passed.

  - **Root Cause**:
    - The first INF appears at:
        - `__module.encoder.block.1.layer.1.DenseReluDense.wo/aten::linear/MatMul`
        - prefix dump: `INF_COUNT=5`, `FINITE_MAX=60928`
    - Those INF values propagate through residual add and trigger NaN at downstream RMS (`INF -> rsqrt(INF)=0 -> INF*0=NaN`).
    - The failure is caused by FP16 precision-path policy on this chain, not by an isolated RMS kernel implementation bug.

  - **Resolution**:
    - Added `MarkRMS` MatcherPass inside `MarkSugraphsToKeepInMixedPrecision` (`mark_subgraphs_to_keep_in_mixed_precision.cpp`).
    - The pass matches all `ov::op::internal::RMS` nodes via `wrap_type<RMS>()` and marks them with `disable_conversion(node, f16)`.
    - Intentionally do not force all MatMul ops to FP32: that would over-broaden precision changes and risk unnecessary performance impact, while the required fix is to enforce the downstream RMS FP32 contract and let mixed-precision alignment choose equivalent valid forms.
    - Nodes already marked by other passes (e.g. `DisableFP16CompForGemma3RMSPattern`) are skipped.
    - This triggers automatically when `keep_precision_sensitive_in_fp32=true` is set in `ConvertPrecision`, which is the standard GPU plugin configuration.
    - RMS is element-wise and memory-bound — keeping it in FP32 has negligible performance impact across all models.

#### Reproduction step and snapshot (if applicable. Do not attach for customer model)
- python -m pytest test_ovc_mo.py \
    --tb=native \
    --env_conf=.automation/env_config.yml \
    --test_conf=.automation/test_configs/desktop_test_config.yml \
    -m "not launch_only_if_manually_specified" \
    --pregen_irs=irs_mapping.csv \
    --tf_models_version=1.15.2 \
    --modules pipelines/production/pytorch/heavy/parler_tts.py \
    -k "Pytorch_ParlerTTSMini_batch_1_device_GPU_precision_FP16" \
    --dynamism_type=None \
    --log-cli-level INFO

#### Problematic graph
 - Original IR
image
- Current IR
image

#### Checklist
 - [x] Is it a proper fix? (not a workaround)
 - [x] Did you include test case for this fix, if necessary?
 - [x] Did you review existing test that can be extended to cover this scenario? Which test did you review?

### Tickets:
  - *CVS-187992*

### CODE REVIEW

File: src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp [hunk 1]
```diff
@@ -136,3 +138,40 @@ TEST(TransformationTests, DisableFP16CompForRMS_Negative) {
...
+            found_rms = true;
+            ASSERT_TRUE(ov::is_conversion_disabled(op, ov::element::f16))
+                << "RMS node must have FP16 conversion disabled to prevent NaN from overflow";
+        }
+        if (op->get_friendly_name() == "matmul_consumer") {
+            found_matmul = true;
+            ASSERT_FALSE(ov::is_conversion_disabled(op, ov::element::f16))
+                << "MatMul consumer should NOT be marked - only RMS needs FP32 protection";
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp]: Updated

File: src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_rms_norm_block.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        auto rms = ov::as_type_ptr<ov::op::internal::RMS>(node);
+        if (!rms)
+            continue;
+
+        if (ov::is_conversion_disabled(rms, ov::element::f16))
+            continue;
+
+        ov::disable_conversion(rms, ov::element::f16);
+        is_changed = true;
+    }
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_rms_norm_block.cpp]: Updated.

File: src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp [hunk 3]
```diff
@@ -136,3 +138,34 @@ TEST(TransformationTests, DisableFP16CompForRMS_Negative) {
...
+        if (op->get_friendly_name() == "rms_target") {
+            ASSERT_TRUE(ov::is_conversion_disabled(op, ov::element::f16))
+                << "RMS node must have FP16 conversion disabled to prevent NaN from overflow";
+        }
+        if (op->get_friendly_name() == "matmul_consumer") {
+            ASSERT_FALSE(ov::is_conversion_disabled(op, ov::element::f16))
+                << "MatMul consumer should NOT be marked - only RMS needs FP32 protection";
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp]: Updated.

File: src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp [hunk 4]
```diff
@@ -136,3 +138,49 @@ TEST(TransformationTests, DisableFP16CompForRMS_Negative) {
...
+    auto input = std::make_shared<ov::op::v0::Parameter>(ov::element::f32, ov::PartialShape{1, 39, 768});
+    auto gamma = ov::op::v0::Constant::create(ov::element::f32, ov::Shape{768}, {1.0f});
+    auto rms = std::make_shared<ov::op::internal::RMS>(input, gamma, 1e-6);
+    rms->set_friendly_name("rms_target");
+
+    auto weight = ov::op::v0::Constant::create(ov::element::f32, ov::Shape{768, 768}, {0.01f});
+    auto matmul = std::make_shared<ov::op::v0::MatMul>(rms, weight, false, true);
+    matmul->set_friendly_name("matmul_consumer");
+
+    auto model = std::make_shared<ov::Model>(ov::OutputVector{matmul}, ov::ParameterVector{input});
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp [hunk 5]
```diff
@@ -136,3 +138,51 @@ TEST(TransformationTests, DisableFP16CompForRMS_Negative) {
...
+    bool found_rms = false;
+    bool found_matmul = false;
+    for (const auto& op : model->get_ops()) {
+        if (op->get_friendly_name() == "matmul_producer") {
+            found_matmul = true;
+            ASSERT_FALSE(ov::is_conversion_disabled(op, ov::element::f16))
+                << "MatMul producer should NOT be marked - it should convert to FP16 normally";
+            ASSERT_EQ(op->get_output_element_type(0), ov::element::f16)
+                << "MatMul should be converted to FP16";
+        }
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp]: Done

File: src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_rms_norm_block.hpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * nodes that already have the disable_conversion mark set.
+ *
+ * Performance impact is negligible: RMS is element-wise and memory-bound,
+ * not compute-bound like MatMul.
+ */
+class DisableFP16CompForRMSNormBlock : public ov::pass::ModelPass {
+public:
+    OPENVINO_MODEL_PASS_RTTI("DisableFP16CompForRMSNormBlock");
+    bool run_on_model(const std::shared_ptr<ov::Model>& model) override;
+};
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_rms_norm_block.hpp]: Done

File: src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.hpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2018-2026 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/pass.hpp"
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.hpp]: Done.

[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.hpp]: Done.

File: src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp [hunk 8]
```diff
@@ -136,3 +139,67 @@ TEST(TransformationTests, DisableFP16CompForRMS_Negative) {
...
+
+                const auto upstream = convert->get_input_node_shared_ptr(0);
+                ASSERT_EQ(upstream->get_friendly_name(), "matmul_producer")
+                    << "Convert feeding RMS must come from matmul_producer";
+            } else {
+                ASSERT_EQ(producer->get_friendly_name(), "matmul_producer")
+                    << "RMS input producer must be matmul_producer when no Convert is inserted";
+                ASSERT_EQ(producer->get_output_element_type(0), ov::element::f32)
+                    << "Without Convert, matmul_producer must stay in f32";
+            }
```
[the remark this answers is not in the document]
Comment (zhanmyz) [reply src/plugins/intel_gpu/tests/unit/transformations/disable_fp16_compression_rms_test.cpp]: This pass guarantees RMS stays precision-protected and receives FP32 input.
Whether that FP32 path is represented by an explicit Convert node or by backend/decompression-aligned rewriting is an implementation detail of the broader mixed-precision pipeline.
In our latest evidence from execute graph/dumps, the fixed path is often materialized as *_decompressed_to_f32 rather than a standalone Convert node, while still satisfying the required numerical contract and removing the INF/NaN issue.
So we keep this test permissive for equivalent valid paths and avoid over-constraining to a single IR form.

File: src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "openvino/core/rt_info.hpp"
+#include "ov_ops/rms.hpp"
+#include "transformations/rt_info/disable_precision_conversion.hpp"
+
+namespace ov::intel_gpu {
+
+DisableFP16CompForAllRMS::DisableFP16CompForAllRMS() {}
+
+bool DisableFP16CompForAllRMS::run_on_model(const std::shared_ptr<ov::Model>& model) {
```
Comment (CuriousPanCake) [inline src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.cpp]: This needs to be detected via MatcherPass

**Comment by wilson-seok** [comment]
As @CuriousPanCake review, please implement this via PatternMatcher.

File: src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.cpp (hunk 9 above)
Comment (zhanmyz) [reply src/plugins/intel_gpu/src/plugin/transformations/disable_fp16_comp_for_all_rms.cpp]: Thanks for the review. Done. Refactored to MatcherPass with `wrap_type<ov::op::internal::RMS>()` pattern matching.

**Comment by zhanmyz** [comment]
> As @CuriousPanCake review, please implement this via PatternMatcher.

Sure, refactored the implementation.

**Comment by CuriousPanCake** [comment]
Shouldn't this be a part of the ConvertPrecision transformation where we mark sensitive nodes to keep them in high precision?

**Review by mryzhov — CHANGES_REQUESTED** [review]
We have a special transformations for such cases, please extend them instead of introducing the new one: src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp

**Comment by zhanmyz** [comment]
> Shouldn't this be a part of the ConvertPrecision transformation where we mark sensitive nodes to keep them in high precision?

You're right. I've moved the fix into `MarkSugraphsToKeepInMixedPrecision` (inside `mark_subgraphs_to_keep_in_mixed_precision.cpp`) as a `MarkRMS` pass, alongside `MarkDivWithEps`, `MarkExp`, and `MarkRandomUniform`. The separate GPU-only pass has been removed.

**Comment by zhanmyz** [comment]
> We have a special transformations for such cases, please extend them instead of introducing the new one: src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp

Done. Added `MarkRMS` class inside `mark_subgraphs_to_keep_in_mixed_precision.cpp` and registered it in `MarkSugraphsToKeepInMixedPrecision::run_on_model`. The standalone GPU pass (`disable_fp16_comp_for_all_rms.cpp/.hpp`) has been deleted. Unit test moved to `mark_subgraph_to_keep_in_mixed_precision_test.cpp` as well.

File: src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp [hunk 10]
```diff
@@ -399,6 +400,29 @@ class MarkDivWithEps : public MatcherPass {
...
+        MATCHER_SCOPE(MarkRMS);
+        auto rms_pattern = pattern::wrap_type<ov::op::internal::RMS>();
+
+        matcher_pass_callback callback = [=](pattern::Matcher& m) {
+            const auto& node = m.get_match_root();
+            if (!node)
+                return false;
+
+            if (is_conversion_disabled(node, element::f16))
+                return false;
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp]: Can we make this check a part of the pattern using a predicate? This way we would avoid entering the callback at all.

File: src/common/transformations/tests/common_optimizations/mark_subgraph_to_keep_in_mixed_precision_test.cpp [hunk 11]
```diff
@@ -1384,4 +1385,44 @@ TEST_F(TransformationTestsF, MarkRandomUniformAsPrecisionSensitive) {
...
+
+        disable_conversion(rms, element::f16);
+        disable_conversion(gamma, element::f16);
+
+        model_ref = make_shared<Model>(OutputVector{rms}, ParameterVector{input});
+    }
+
+    const FunctionsComparator func_comparator =
+        FunctionsComparator::with_default().enable(FunctionsComparator::RUNTIME_KEYS);
+    // need to compare twice to ensure that no extra nodes are marked
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/mark_subgraph_to_keep_in_mixed_precision_test.cpp:1423]: Why?

File: src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp (hunk 10 above)
Comment (zhanmyz) [reply src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp]: Good point. Moved the `is_conversion_disabled` check into the pattern predicate so nodes already marked won't enter the callback at all. Updated.

File: src/common/transformations/tests/common_optimizations/mark_subgraph_to_keep_in_mixed_precision_test.cpp (hunk 11 above)
Comment (zhanmyz) [reply src/common/transformations/tests/common_optimizations/mark_subgraph_to_keep_in_mixed_precision_test.cpp:1423]: This is the standard pattern used by all tests in this file (e.g. `MarkDivWithEps`, etc.). The first comparison (`model_ref` vs `model`) verifies all expected marks are present. The second comparison (`model` vs `model_ref`) verifies no extra nodes were incorrectly marked.

Comment (CuriousPanCake) [reply src/common/transformations/tests/common_optimizations/mark_subgraph_to_keep_in_mixed_precision_test.cpp:1423]: I see.

**Review by CuriousPanCake — APPROVED** [review]
LGTM

**Review by mryzhov — APPROVED** [review]

**Review by e-ddykim — APPROVED** [review]
no perf. impact from llm daily tests
